subprojects/core/src/main/java/org/gradle/api/internal/project/MutableStateAccessAwareProject.java

```diff
@@ -0,0 +1,1088 @@
+        onMutableStateAccess("components");
+        delegate.components(configuration);
+    }
+
+    @Override
+    public ProjectConfigurationActionContainer getConfigurationActions() {
+        return delegate.getConfigurationActions();
+    }
+
+    @Override
+    public ModelRegistry getModelRegistry() {
+        return delegate.getModelRegistry();
+    }
+
+    @Override
+    public ClassLoaderScope getClassLoaderScope() {
+        return delegate.getClassLoaderScope();
+    }
+
+    @Override
+    public ClassLoaderScope getBaseClassLoaderScope() {
+        return delegate.getBaseClassLoaderScope();
+    }
+
+    @Override
+    public void setScript(Script script) {
+        delegate.setScript(script);
+    }
+
+    @Override
+    public void addDeferredConfiguration(Runnable configuration) {
+        delegate.addDeferredConfiguration(configuration);
+    }
+
+    @Override
+    public void fireDeferredConfiguration() {
+        delegate.fireDeferredConfiguration();
+    }
+
+    @Override
+    public Path identityPath(String name) {
+        return delegate.identityPath(name);
+    }
+
+    @Override
+    public Path projectPath(String name) {
+        return delegate.projectPath(name);
+    }
+
+    @Nonnull
+    @Override
+    public Path getProjectPath() {
+        return delegate.getProjectPath();
+    }
+
+    @Override
+    public ProjectInternal getProject() {
+        return this;
+    }
+
+    @Override
+    public ModelContainer<?> getModel() {
+        return delegate.getModel();
+    }
+
+    @Override
+    public Path getBuildPath() {
+        return delegate.getBuildPath();
+    }
+
+    @Override
+    public boolean isScript() {
+        return delegate.isScript();
+    }
+
+    @Override
+    public boolean isRootScript() {
+        return delegate.isRootScript();
+    }
+
+    @Override
+    public boolean isPluginContext() {
+        return delegate.isPluginContext();
+    }
+
+    @Override
+    public Path getIdentityPath() {
+        return delegate.getIdentityPath();
+    }
+
+    @Nullable
+    @Override
+    public ProjectEvaluationListener stepEvaluationListener(ProjectEvaluationListener listener, Action<ProjectEvaluationListener> action) {
+        return delegate.stepEvaluationListener(listener, action);
+    }
+
+    @Override
+    public ProjectState getOwner() {
+        return delegate.getOwner();
+    }
+
+    @Override
+    public InputNormalizationHandlerInternal getNormalization() {
+        return delegate.getNormalization();
+    }
+
+    @Override
+    public void normalization(Action<? super InputNormalizationHandler> configuration) {
+        delegate.normalization(configuration);
+    }
+
+    @Override
+    public void dependencyLocking(Action<? super DependencyLockingHandler> configuration) {
+        delegate.dependencyLocking(configuration);
+    }
+
+    @Override
+    public DependencyLockingHandler getDependencyLocking() {
```

Comment:
❌ This is mutable state as well

Reply:
Taking a look at implementation details of the service - I couldn't understand how the state of the service could be mutated. But then I found that `DependencyLockingProvider` has a bunch of state. Could you confirm that `DependencyLockingProvider` is a source of possible state mutations from project level?